From a5ba3d3c4dd5c63241dac4208af3775daf0a7f08 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Mon, 24 Aug 2015 20:39:29 +0530
Subject: [PATCH] [EM/PMIC_CCSM]: Fix for USB Charger type detection

Some times the charger is not getting detected properly, to make sure
the charger type is detected by pmic ccsm re-starting internal charger
type detection on usb port.

Change-Id: I89f1d14398ff96ad780064a74de8480b88057ac4
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-14897
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
---
 drivers/power/intel_pmic_ccsm.c | 10 +++++++---
 drivers/power/intel_pmic_ccsm.h |  5 ++++-
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/power/intel_pmic_ccsm.c b/drivers/power/intel_pmic_ccsm.c
index 6ac692268d94..8eaedb0c7f54 100644
--- a/drivers/power/intel_pmic_ccsm.c
+++ b/drivers/power/intel_pmic_ccsm.c
@@ -857,14 +857,18 @@ static int get_charger_type(void)
 		if ((val & USBSRCDET_SUSBHWDET_DETSUCC) ==
 				USBSRCDET_SUSBHWDET_DETSUCC)
 			break;
-		else
-			msleep(USBSRCDET_SLEEP_TIME);
+		else {
+			pmic_write_reg(chc.reg_map->pmic_usbphyctrl,
+						USBPHYCTRL_CTYPE_START);
+			usleep_range(USBSRCDET_SLEEP_MIN_TIME,
+					USBSRCDET_SLEEP_MAX_TIME);
+		}
 	} while (i < USBSRCDET_RETRY_CNT);
 
 	if ((val & USBSRCDET_SUSBHWDET_DETSUCC) !=
 			USBSRCDET_SUSBHWDET_DETSUCC) {
 		dev_err(chc.dev, "Charger detection unsuccessful after %dms\n",
-			i * USBSRCDET_SLEEP_TIME);
+			i * USBSRCDET_SLEEP_MIN_TIME);
 		return 0;
 	}
 
diff --git a/drivers/power/intel_pmic_ccsm.h b/drivers/power/intel_pmic_ccsm.h
index 515901133162..9e193ae1b302 100644
--- a/drivers/power/intel_pmic_ccsm.h
+++ b/drivers/power/intel_pmic_ccsm.h
@@ -149,7 +149,8 @@
 
 
 #define USBSRCDET_RETRY_CNT		5
-#define USBSRCDET_SLEEP_TIME		200
+#define USBSRCDET_SLEEP_MIN_TIME	300000
+#define USBSRCDET_SLEEP_MAX_TIME	900000
 #define USBSRCDET_SUSBHWDET_MASK	(3 << 0)
 #define USBSRCDET_USBSRCRSLT_MASK	(0x0F << 2)
 #define USBSRCDET_SDCD_MASK		(1 << 6)
@@ -160,6 +161,8 @@
 #define USBPHYCTRL_CHGDET_N_POL_MASK	(1 << 1)
 #define USBPHYCTRL_USBPHYRSTB_MASK	(1 << 0)
 
+#define USBPHYCTRL_CTYPE_START		(1 << 2)
+
 /* Registers on I2C-dev2-0x6E */
 #define USBPATH_USBSEL_MASK	(1 << 3)
 
-- 
1.9.1

